Configure git to fetch private go repositories

moorara opened the pull request:
Related to https://github.com/renovatebot/renovate/issues/3202

This PR adds an `entrypoint` to Docker image for `go`. If a `GITHUB_TOKEN` environment variable is set/passed, this script will add a configuration to git client, so *private repositories* on GitHub can be fetched.

I see two concerns:

  - [x] Handle _GitHub_ 
  - [ ] Support _GitHub Enterprise_
  - [x] Make this change non-breaking (renovate passes `go` as docker command currently)

rarkins commented:
@moorara thanks for this PR to move things along. I'm trying to think up the advantages and disadvantages of various approaches to this problem:

Approach 1: make no changes to the image itself
- In this approach we'd run like `docker run --rm -e GITHUB_TOKEN renovate/go bash -c "git config --global url.\"https://$GITHUB_TOKEN@github.com/\".insteadOf \"https://github.com/\" && go get`
- Advantage: we can easily add as many different tokens as we need (e.g. GitHub.com, GitHub Enterprise, and others)
- Disadvantage: the run string looks pretty long

Approach 2: Pre-configure a git config file and then just supply the tokens in env
- Advantage: seems cleaner
- Disadvantage: although it would work cleanly for github.com right now, it would be the hardest approach for managing multiple/optional host authentications

Approach 3: This PR with a custom entrypoint to detect env and then set it
- Advantage: It makes the `go get` command from Renovate look normal, it handles optionality of the token
- Disadvantage: I can't see how we can scale this to multiple and custom platforms in future, e.g. for https://github.mycompany.com etc. Also we need to overcome the backwards compatibility problem using either a more advanced entrypoint script or a custom tag.

moorara commented:
@rarkins @emman27
I updated the PR two address the aforementioned concerns.

To make this approach working with multiple git provider, we can add a check for each type of provider in the `entrypoint.sh`. I added two for **bitbucket.org** and **gitlab.com**.

To make the docker image backward-compatible, we check the set/passed arguments in`entrypoint.sh`.

emman27 commented:
Is there some way we can get this extensible to Github Enterprise users?

Say, if it's just 3 of us each owning our own domains, is there some way we could possibly support the following domains?

git.moorara.com
github.emman27.com
vcs.rarkins.org

moorara commented:
@emman27 I am not really familiar with on-prem setup of GitHub or other VCS.
So, I don't know if every deployment of GitHub Enterprise allow token-based authentication for example. Any idea?

moorara commented:
@rarkins @emman27 
Can we come up with a design for GitHub Enterprise and unblock the case of GitHub for now?

Not being able to fetch private go modules is blocking us and prevent our organization to use Renovate since all of our Renovate PRs are now failing.

emman27 commented:
True, it's blocking me as well. I'd love to see this merged as a temp fix, but hope that there's going to be more to support such use cases in the future!

rarkins commented:
I plan to fix this on a high priority in the next week, but first I need a way to "reproduce" the problem. If someone can create two sample repositories - one to be the source module (to be put in a private repo) and the other to be the module consumer - then it would certainly help me save some time. Otherwise, please stay tuned for updates.

moorara commented:
@rarkins 

I created these two repos:

  - **private go module:** https://github.com/moorara/test-go
  - **consumer repo:**: https://github.com/moorara/test-service

On the consumer repo, there is a renovate PR failing.

rarkins commented:
Thanks @moorara! Could you temporarily switch test-go to be public? I'll then clone it for myself to use/test against first. After that we can switch it back to private and use it as the first repo to test the bot against once the feature is ready.

moorara commented:
@rarkins It's now public. Thanks.

rarkins commented:
@moorara thanks, this was very helpful for verifying my code.

I've decided to internalize this auth into Renovate itself rather than this Go package, so this PR can now be closed. Support is in `renovate@15.3.0`, which is live in the app.

Summarizing:

- Right now, only github.com auth is supported but I'm going to add some issues to track adding other hosts
- In future, the three main public code hosts will be supported automatically without further configuration: github.com, gitlab.com and bitbucket.org
- In future, any self-hosted source host (e.g. GitHub Enterprise, Bitbucket Server, GitLab CE/EE) will also get auth passed automatically without further configuration
- Any other host will be added manually by setting a host rule with platform=go
